lib/app.rb

```diff
@@ -192,35 +192,47 @@ def query_string_for_redirection
       request.query_string.empty? ? nil : "?#{request.query_string}"
     end
 
-    def manifest_asset_urls
-      @@manifest_asset_urls ||= [
+    def service_worker_asset_urls
+      @@service_worker_asset_urls ||= [
         javascript_path('application', asset_host: false),
         stylesheet_path('application'),
         image_path('docs-1.png'),
         image_path('docs-1@2x.png'),
         image_path('docs-2.png'),
         image_path('docs-2@2x.png'),
-        asset_path('docs.js')
-      ]
+        asset_path('docs.js'),
+        App.production? ? nil : javascript_path('debug'),
+      ].compact
     end
 
-    def app_size
-      @app_size ||= memoized_cookies['size'].nil? ? '20rem' : "#{memoized_cookies['size']}px"
-    end
+    # Returns a cache name for the service worker to use which changes if any of the assets changes
+    # When a manifest exist, this name is only created once based on the asset manifest because it never changes without a server restart
+    # If a manifest does not exist, it is created every time this method is called because the assets can change while the server is running
+    def service_worker_cache_name
+      if File.exist?(App.assets_manifest_path)
+        if defined?(@@service_worker_cache_name)
+          return @@service_worker_cache_name
+        end
 
-    def app_layout
-      memoized_cookies['layout']
-    end
+        digest = Sprockets::Manifest
+                   .new(nil, App.assets_manifest_path)
+                   .files
+                   .values
+                   .map {|file| file["digest"]}
+                   .join
 
-    def app_theme
-      @app_theme ||= memoized_cookies['dark'].nil? ? 'default' : 'dark'
-    end
+        return @@service_worker_cache_name ||= Digest::MD5.hexdigest(digest)
+      else
+        paths = App.sprockets
+                  .each_file
+                  .to_a
+                  .reject {|file| file.start_with?(App.docs_path)}
 
-    def dark_theme?
-      app_theme == 'dark'
+        return App.sprockets.pack_hexdigest(App.sprockets.files_digest(paths))
+      end
     end
 
-    def redirect_via_js(path) # courtesy of HTML5 App Cache
```

Comment:
Is this redirect still necessary with the Service Worker?

Reply:
Seems like it was never really necessary. All it does is ensure the correct page is loaded when you directly navigate to a page like [https://devdocs.io/offline](https://devdocs.io/offline) (instead of using the dropdown menu). The `redirect_via_js` method redirects the user to `https://devdocs.io/` and lets the JS code know which page to show through a cookie. The alternative that is used if the user has a browser that doesn't support this is to send the user to `https://devdocs.io/#/offline`, while also letting the JS code handle the routing.

It seems like the `redirect_via_js` method was added in 217d5016d05ecb15a872139336fe2453adab70d6 because the other approach seemed to cause a bug when directly navigating to certain documentation entries. I have no idea if this bug still exists, nor do I know how to test if it does because of the vague commit message, so I think it's best to keep the method in there.